source/common/protobuf/message_converter.cc

```diff
@@ -0,0 +1,155 @@
+#include "source/common/protobuf/message_converter.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "source/common/buffer/buffer_impl.h"
+
+#include "absl/log/absl_check.h"
+#include "google/protobuf/io/zero_copy_stream.h"
+#include "grpc_transcoding/message_reader.h"
+
+namespace Envoy::ProtobufMessage {
+
+namespace {
+using Envoy::Protobuf::io::ZeroCopyInputStream;
+using ::google::grpc::transcoding::kGrpcDelimiterByteSize;
+using ::google::grpc::transcoding::MessageReader;
+} // namespace
+
+MessageConverter::MessageConverter(std::unique_ptr<CreateMessageDataFunc> factory,
+                                   uint32_t buffer_limit)
+    : factory_(std::move(factory)), buffer_limit_(buffer_limit),
+      parsed_bytes_usage_(std::make_shared<uint64_t>(0)) {}
+
+absl::StatusOr<std::unique_ptr<StreamMessage>>
+MessageConverter::AccumulateMessage(Envoy::Buffer::Instance& data, bool end_stream) {
+  // Append the input data to buffer that will be parsed.
+  parsing_buffer_.move(data);
+
+  ENVOY_LOG_MISC(info, "Checking buffer limits: actual {} > limit {}?", BytesBuffered(),
+                 buffer_limit_);
+  if (BytesBuffered() > buffer_limit_) {
+    return absl::FailedPreconditionError("Rejected because internal buffer limits are exceeded.");
+  }
+
+  auto parsed_output = MaybeParseGrpcMessage(*factory_, parsing_buffer_);
```

Comment:
nit: Please use the explicit type here.

https://google.github.io/styleguide/cppguide.html#Type_deduction

Reply:
Done.